pkg/execution/executor/service.go

```diff
@@ -114,47 +116,110 @@ func (s *svc) Executor() execution.Executor {
 	return s.exec
 }
 
-func (s *svc) getFailureHandler(ctx context.Context) (func(context.Context, state.Identifier, state.State, state.DriverResponse) error, error) {
+func (s *svc) getFinishHandler(ctx context.Context) (func(context.Context, state.Identifier, state.State, state.DriverResponse) error, error) {
 	pb, err := pubsub.NewPublisher(ctx, s.config.EventStream.Service)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create publisher: %w", err)
 	}
 
 	topicName := s.config.EventStream.Service.Concrete.TopicName()
 
-	return func(ctx context.Context, id state.Identifier, s state.State, r state.DriverResponse) error {
+	return func(ctx context.Context, id state.Identifier, st state.State, r state.DriverResponse) error {
+		eg := errgroup.Group{}
 		now := time.Now()
-		evt := event.Event{
-			ID:        ulid.MustNew(uint64(now.UnixMilli()), rand.Reader).String(),
-			Name:      event.FnFailedName,
-			Timestamp: now.UnixMilli(),
-			Data: map[string]interface{}{
-				"function_id": s.Function().Slug,
-				"run_id":      id.RunID.String(),
-				"error":       r.UserError(),
-				"event":       s.Event(),
-			},
-		}
 
-		byt, err := json.Marshal(evt)
-		if err != nil {
-			return fmt.Errorf("error marshalling failure event: %w", err)
-		}
+		// Legacy - send inngest/function.failed
+		if r.Err != nil {
+			eg.Go(func() error {
+				evt := event.Event{
+					ID:        ulid.MustNew(uint64(now.UnixMilli()), rand.Reader).String(),
+					Name:      event.FnFailedName,
+					Timestamp: now.UnixMilli(),
+					Data: map[string]interface{}{
+						"function_id": st.Function().Slug,
+						"run_id":      id.RunID.String(),
+						"error":       r.UserError(),
+						"event":       st.Event(),
+					},
+				}
 
-		err = pb.Publish(
-			ctx,
-			topicName,
-			pubsub.Message{
-				Name:      event.EventReceivedName,
-				Data:      string(byt),
-				Timestamp: now,
-			},
-		)
-		if err != nil {
-			return fmt.Errorf("error publishing failure event: %w", err)
+				byt, err := json.Marshal(evt)
+				if err != nil {
+					return fmt.Errorf("error marshalling failure event: %w", err)
+				}
+
+				err = pb.Publish(
+					ctx,
+					topicName,
+					pubsub.Message{
+						Name:      event.EventReceivedName,
+						Data:      string(byt),
+						Timestamp: now,
+					},
+				)
+				if err != nil {
+					return fmt.Errorf("error publishing failure event: %w", err)
+				}
+
+				return nil
+			})
 		}
 
-		return nil
+		// send inngest/function.finished
```

Comment:
Are we doing this on _every_ function finish now?  Or only if there's a correlation ID?

Additionally, should the "function finished" event generation be bundled into the executor itself?  In the cloud we obviously have a wrapper such, but the actual generation of the `event.Event` is the same, right?

Reply:
@tonyhb That's the plan atm from https://github.com/inngest/inngest/pull/782#discussion_r1376256525. We can limit, though an `inngest/function.finished` event is pretty cool to be able to listen to across the system.

It doubles the number of events, of course...